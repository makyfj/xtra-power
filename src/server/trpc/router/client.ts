import { z } from "zod";
import { TRPCError } from "@trpc/server";

import { t, authedProcedure } from "../trpc";

export const clientRouter = t.router({
  createClient: authedProcedure
    .input(
      z.object({
        name: z.string().nullable(),
        email: z.string().nullable(),
        image: z.string().nullable(),
        userId: z.string(),
      })
    )
    .mutation(async ({ ctx, input: { name, email, image, userId } }) => {
      const clientExist = await ctx.prisma.client.findUnique({
        where: {
          email: email as string,
        },
      });

      if (clientExist) {
        return await ctx.prisma.user.update({
          where: {
            id: userId,
          },
          data: {
            clientId: clientExist.id,
          },
        });
      }

      const client = await ctx.prisma.client.create({
        data: {
          name,
          email,
          image,
          userId,
        },
      });

      if (!client) {
        throw new TRPCError({
          code: "INTERNAL_SERVER_ERROR",
          message: "Failed to create client",
        });
      } else {
        // Update the user's client Id
        await ctx.prisma.user.update({
          where: {
            id: userId,
          },
          data: {
            clientId: client.id,
          },
        });

        return client;
      }
    }),

  addTrainer: authedProcedure
    .input(
      z.object({
        userId: z.string(),
        trainerId: z.string(),
      })
    )
    .mutation(({ ctx, input: { userId, trainerId } }) => {
      const clientUpdate = ctx.prisma.client.update({
        where: {
          userId: userId,
        },
        data: {
          trainerId: trainerId,
        },
      });

      const trainerUpdate = ctx.prisma.trainer.update({
        where: {
          id: trainerId,
        },
        data: {
          clients: {
            connect: {
              userId: userId,
            },
          },
        },
      });

      return Promise.all([clientUpdate, trainerUpdate]);
    }),

  getTrainer: authedProcedure
    .input(
      z.object({
        email: z.string(),
      })
    )
    .query(async ({ ctx, input: { email } }) => {
      const client = await ctx.prisma.client.findUnique({
        where: {
          email,
        },
      });

      if (client && client.trainerId) {
        const trainer = await ctx.prisma.trainer.findUnique({
          where: {
            id: client.trainerId,
          },
        });

        return trainer;
      }
    }),

  deleteClient: authedProcedure
    .input(
      z.object({
        email: z.string(),
      })
    )
    .mutation(({ ctx, input: { email } }) => {
      return ctx.prisma.client.delete({
        where: {
          email,
        },
      });
    }),
});
