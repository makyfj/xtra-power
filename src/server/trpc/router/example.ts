import { t } from "../trpc";
import { z } from "zod";

export const exampleRouter = t.router({
  hello: t.procedure
    .input(z.object({ text: z.string() }))
    .query(({ input }) => {
      return {
        greeting: `Hello ${input?.text ?? "world"}`,
        id: "123",
      };
    }),
  getAll: t.procedure.query(({ ctx }) => {
    return ctx.prisma.example.findMany();
  }),
  getById: t.procedure
    .input(z.object({ id: z.string() }))
    .query(({ ctx, input }) => {
      return ctx.prisma.example.findUnique({
        where: {
          id: input.id,
        },
      });
    }),
});
