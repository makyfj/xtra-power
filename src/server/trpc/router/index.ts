// src/server/trpc/router/index.ts
import { t } from "../trpc";
import { exampleRouter } from "./example";
import { authRouter } from "./auth";
import { workoutRouter } from "./workout";

export const appRouter = t.router({
  example: exampleRouter,
  auth: authRouter,
  workout: workoutRouter,
});

// export type definition of API
export type AppRouter = typeof appRouter;
