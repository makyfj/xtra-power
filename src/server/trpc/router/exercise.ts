import { z } from "zod";

import { t, authedProcedure } from "../trpc";

export const exerciseRouter = t.router({
  getExerciseById: authedProcedure
    .input(z.object({ id: z.string() }))
    .query(({ ctx, input }) => {
      return ctx.prisma.exercise.findUnique({
        where: {
          id: input.id,
        },
      });
    }),

  updateExercise: authedProcedure
    .input(
      z.object({ id: z.string(), name: z.string(), description: z.string() })
    )
    .mutation(({ ctx, input }) => {
      return ctx.prisma.exercise.update({
        where: {
          id: input.id,
        },
        data: {
          name: input.name,
          description: input.description,
        },
      });
    }),

  deleteExerciseById: authedProcedure
    .input(z.object({ id: z.string() }))
    .mutation(({ ctx, input }) => {
      return ctx.prisma.exercise.delete({
        where: {
          id: input.id,
        },
      });
    }),

  createExercise: authedProcedure
    .input(
      z.object({
        name: z.string(),
        description: z.string(),
        workoutId: z.string(),
      })
    )
    .mutation(({ ctx, input }) => {
      return ctx.prisma.exercise.create({
        data: {
          name: input.name,
          description: input.description,
          workoutId: input.workoutId,
        },
      });
    }),
});
