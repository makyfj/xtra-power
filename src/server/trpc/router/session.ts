import { z } from "zod";
import { TRPCError } from "@trpc/server";

import { t } from "../trpc";

export const sessionRouter = t.router({
  getSession: t.procedure
    .input(
      z.object({
        email: z.string().nullish(),
      })
    )
    .query(({ ctx, input: { email } }) => {
      if (email) {
        return ctx.session;
      } else {
        throw new TRPCError({
          code: "INTERNAL_SERVER_ERROR",
          message: "Failed to find user",
        });
      }
    }),
});
