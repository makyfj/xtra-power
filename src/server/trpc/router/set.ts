import { t, authedProcedure } from "../trpc";
import { z } from "zod";

export const setRouter = t.router({
  getSetsByExerciseId: authedProcedure
    .input(
      z.object({
        exerciseId: z.string(),
      })
    )
    .query(({ ctx, input }) => {
      return ctx.prisma.set.findMany({
        where: {
          exerciseId: input.exerciseId,
        },
      });
    }),
  createSets: authedProcedure
    .input(
      z.object({
        exerciseId: z.string(),
        sets: z.array(
          z.object({
            weight: z.number(),
            reps: z.number(),
          })
        ),
      })
    )
    .mutation(({ ctx, input: { exerciseId, sets } }) => {
      return ctx.prisma.set.createMany({
        data: sets.map((set) => ({
          exerciseId,
          weight: set.weight,
          reps: set.reps,
        })),
      });
    }),
  updateSets: authedProcedure
    .input(
      z.object({
        sets: z.array(
          z.object({
            id: z.string(),
            weight: z.number(),
            reps: z.number(),
          })
        ),
      })
    )
    .mutation(({ ctx, input: { sets } }) => {
      return ctx.prisma.set.updateMany({
        data: sets.map((set) => ({
          weight: set.weight,
          reps: set.reps,
        })),
        where: {
          id: {
            in: sets.map((set) => set.id),
          },
        },
      });
    }),
  deleteSetById: authedProcedure
    .input(
      z.object({
        id: z.string(),
      })
    )
    .mutation(({ ctx, input: { id } }) => {
      return ctx.prisma.set.delete({
        where: {
          id,
        },
      });
    }),
});
