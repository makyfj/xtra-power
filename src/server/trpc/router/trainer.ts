import { z } from "zod";
import { TRPCError } from "@trpc/server";

import { t, authedProcedure } from "../trpc";

export const trainerRouter = t.router({
  createTrainer: authedProcedure
    .input(
      z.object({
        name: z.string().nullish(),
        email: z.string().nullish(),
        image: z.string().nullish(),
        userId: z.string(),
      })
    )
    .mutation(async ({ ctx, input: { name, email, image, userId } }) => {
      const trainer = await ctx.prisma.trainer.create({
        data: {
          name,
          email,
          image,
        },
      });

      if (!trainer) {
        throw new TRPCError({
          code: "INTERNAL_SERVER_ERROR",
          message: "Failed to create trainer",
        });
      } else {
        const trainerUpdated = await ctx.prisma.user.update({
          where: {
            id: userId,
          },
          data: {
            trainerId: trainer.id,
          },
        });

        return trainerUpdated;
      }
    }),

  getTrainers: authedProcedure.query(({ ctx }) => {
    const trainers = ctx.prisma.trainer.findMany({});

    if (!trainers) {
      throw new TRPCError({
        code: "INTERNAL_SERVER_ERROR",
        message: "Failed to get trainers",
      });
    }

    return trainers;
  }),

  getMyClients: authedProcedure
    .input(
      z.object({
        email: z.string().nullable().nullish(),
      })
    )
    .query(async ({ ctx, input: { email } }) => {
      const clients = ctx.prisma.trainer.findMany({
        where: {
          email,
        },
        include: {
          users: true,
        },
      });

      const trainer = await ctx.prisma.trainer.findUnique({
        where: {
          email: email as string,
        },
      });

      // Get the clients that clientId is not null and trainerId is the same as the trainer's id

      const myClients = await ctx.prisma.user.findMany({
        where: {
          clientId: {
            not: null,
          },
          trainerId: trainer?.id,
        },
      });

      return myClients;
    }),

  getClients: authedProcedure.query(({ ctx }) => {
    const clients = ctx.prisma.user.findMany({
      where: {
        clientId: {
          not: null,
        },
      },
    });

    return clients;
  }),
});
