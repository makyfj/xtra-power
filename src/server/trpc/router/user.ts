import { t, authedProcedure } from "../trpc";
import { z } from "zod";

export const userRouter = t.router({
  getUserByEmail: authedProcedure
    .input(
      z.object({
        email: z.string(),
      })
    )
    .query(({ ctx, input: { email } }) => {
      if (email) {
        const user = ctx.prisma.user.findUnique({
          where: {
            email: email,
          },
        });

        return user;
      }
    }),

  getUser: authedProcedure.query(({ ctx }) => {
    const { session, prisma } = ctx;
    if (session && session.user) {
      const email = session.user.email as string;
      return prisma.user.findUnique({
        where: {
          email,
        },
      });
    }
  }),

  createUser: authedProcedure
    .input(
      z.object({
        name: z.string(),
        email: z.string(),
      })
    )
    .mutation(({ ctx, input: { name, email } }) => {
      const user = ctx.prisma.user.create({
        data: {
          name: name,
          email: email,
        },
      });

      return user;
    }),

  updateTypeUser: authedProcedure
    .input(
      z.object({
        id: z.string(),
        clientId: z.string().nullable(),
        trainerId: z.string().nullable(),
      })
    )
    .mutation(({ ctx, input: { id, clientId, trainerId } }) => {
      if (clientId) {
        const updatedUser = ctx.prisma.user.update({
          where: {
            id,
          },
          data: {
            clientId,
          },
        });
        return updatedUser;
      } else {
        const updatedUser = ctx.prisma.user.update({
          where: {
            id,
          },
          data: {
            trainerId,
          },
        });
        return updatedUser;
      }
    }),
  updateUser: authedProcedure
    .input(
      z.object({
        id: z.string(),
        name: z.string(),
      })
    )
    .mutation(({ ctx, input: { id, name } }) => {
      const user = ctx.prisma.user.update({
        where: {
          id,
        },
        data: {
          name,
        },
      });

      return user;
    }),

  deleteUser: authedProcedure
    .input(z.object({ id: z.string() }))
    .mutation(({ ctx, input: { id } }) => {
      const user = ctx.prisma.user.delete({
        where: {
          id,
        },
      });

      return user;
    }),
});
