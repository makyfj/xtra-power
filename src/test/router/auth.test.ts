/**
 * Integration test example for the `auth` router
 */
import { createContextInner } from "src/server/trpc/context";
import { AppRouter, appRouter } from "src/server/trpc/router/index";
import { inferProcedureOutput } from "@trpc/server";

test("Check if user is logged in and get the user's session", async () => {
  const ctx = await createContextInner({
    session: {
      user: {
        id: "1",
      },
      expires: new Date().toDateString(),
    },
  });

  const caller = appRouter.createCaller(ctx);

  const output: inferProcedureOutput<AppRouter["auth"]["getSession"]> = {
    user: {
      id: "1",
    },
    expires: new Date().toDateString(),
  };

  const session = await caller.auth.getSession();

  expect(session).toMatchObject(output);
});

test("Get secret message when logged in", async () => {
  const ctx = await createContextInner({
    session: {
      user: {
        id: "1",
      },
      expires: new Date().toDateString(),
    },
  });

  const caller = appRouter.createCaller(ctx);

  const output: inferProcedureOutput<AppRouter["auth"]["getSecretMessage"]> =
    "You are logged in and can see this secret message!";

  const secretMessage = await caller.auth.getSecretMessage();

  expect(secretMessage).toMatchSnapshot(output)
});
