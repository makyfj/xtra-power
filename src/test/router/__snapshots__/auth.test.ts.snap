// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get secret message when logged in: You are logged in and can see this secret message! 1`] = `"You are logged in and can see this secret message!"`;
